lib/attrsets.nix

```diff
@@ -1798,6 +1799,51 @@ rec {
       then pkg.${output} or pkg.out or pkg
       else pkg;
 
+  /**
+    Get the first of the `outputs` provided by the package, or the default.
+    This function is alligned with `_overrideFirst()` from the `multiple-outputs.sh` setup hook.
+    Like `getOutput`, the function is idempotent.
+
+    # Inputs
+
+    `outputs`
+
+    : 1\. Function argument
+
+    `pkg`
+
+    : 2\. Function argument
+
+    # Type
+
+    ```
+    getFirstOutput :: [String] -> (Derivation | DerivationOutput) -> DerivationOutput
+    ```
+
+    # Examples
+    :::{.example}
+    ## `lib.attrsets.getFirstOutput` usage example
+
+    ```nix
+    getFirstOutput [ "include" "dev" ] pkgs.openssl
+    => "/nix/store/00000000000000000000000000000000-openssl-1.0.1r-dev"
+    ```
+
+    :::
+  */
+  getFirstOutput =
```

Comment:
How about this implementation, which is shorter, doesn’t scan the entire list, and maintains the `out` fallback of `lib.getOutput`?

```
  getFirstOutput = candidates: pkg:
    if pkg.outputSpecified or false then
      pkg
    else
      pkg.${lib.findFirst (builtins.hasAttr pkg) "out" candidates} or pkg
```

Reply:
But do we want to maintain the `out` fallback in the `getFirstOutput` version? You argued before, and I maybe concur, that it's cleaner to leave `out` out, and define `getOutput = output: getFirstOutput [ output "out" ]`, instead of `getOutput = output: getFirstOutput [ output ]` and keeping `out` in the definition

Reply:
`findFirst` looks nice though

Reply:
I guess the question is whether there’s any use‐case where you wouldn’t want `out` in here. Without the `or pkg`, I could imagine circumstances where you might want `null` instead, but if we’re falling back to `pkg` I think there’s no reason you wouldn’t want `pkg.out` if available instead.

Reply:
Here's where we are right now. Derivations that do not

```
nix-repl> pkgs' = lib.filterAttrs (name: value: (builtins.tryEval (lib.isDerivation value && value ? outputs && !(builtins.elem "out" (value.outputs or [])))).value) pkgs
nix-repl> builtins.attrNames pkgs'                                                                                                                                         
[ "sagetex" ]
nix-repl> sagetex.outputs
[ "tex" ]
```

The choice also affects packages that list other outputs before `out`, e.g., with the current implementation:

```
nix-repl> (lib.getFirstOutput [ "lib" "out" ] sqlite).outputName     
"out"
nix-repl> (lib.getFirstOutput [ "lib" ] sqlite).outputName        
"bin"
```

Reply:
I think that the facts that a `pkg` default that doesn’t match the `lib.getOutput` behaviour of checking `out` first seems like a potential footgun, and that `lib.findFirst` dovetails nicely with that default, make me lean towards including it here. But I’m surprised that there’s only one package without an `out` at all!

Reply:
I also like that `getInclude = getFirstOutput [ "include" "dev" "out" ]` lists `out` explicitly, like `_overrideFirst` does

Reply:
>  I’m surprised that there’s only one package without an out at all!

Maybe if I recursed I'd have found more

> a potential footgun

Yes, OTOH it's meant to be used internally to build other `getX` functions

Reply:
I like the `findFirst` solution but I want to keep `getInclude = getFirstOutput [ "include" "dev" "out" ]` as a piece of documentation of sorts